src/Microsoft.AspNetCore.Routing/Tree/TreeRouteBuilder.cs

```diff
@@ -392,5 +393,25 @@ private void AddEntryToTree(UrlMatchingTree tree, InboundRouteEntry entry)
                 return result == 0 ? x.Entry.RouteTemplate.TemplateText.CompareTo(y.Entry.RouteTemplate.TemplateText) : result;
             });
         }
+
+        private static bool RemainingSegmentsAreOptional(IList<TemplateSegment> segments, int currentParameterIndex)
+        {
+            for (var j = currentParameterIndex + 1; j < segments.Count; j++)
+            {
+                if (!(segments[j].IsSimple &&
```

Comment:
weird negation - this consider just breaking this up into a series of `if`s with some comments